### PR TITLE
feat(gemini): implement Google Gemini connector

### LINKED ISSUE #3717: Implement Google Gemini connector
Develop and implement the Google Gemini Connector, enabling integration of advanced AI features into Camunda BPMN workflows. This includes conversational AI, JSON data handling, long context support, and function calling. Ensure secure authentication, robust error handling, and seamless integration.

Related issue

- https://github.com/camunda/product-hub/issues/2543

### PR DESCRIPTION
**Pull request by DenovVasil**
## Description
Implement gemini connector

image

## Related issue
closes 
#3717

### CODE REVIEW

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/caller/GeminiCaller.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+      GenerateContentResponse response = model.generateContent(content);
+
+      return ResponseHandler.getContent(response);
+    }
+  }
+
+  private List<FunctionDeclaration> prepareFunctionDeclarations(List<Object> functionCalls) {
+    List<FunctionDeclaration> result = new ArrayList<>();
+    for (Object functionCall : functionCalls) {
```
Comment (DenovVasil) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/caller/GeminiCaller.java:58]: n

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+        @NotNull
+        List<Object> prompts,
+    @TemplateProperty(
+            label = "System instructions",
+            group = "input",
+            description = "System instructions inform how the model should respond.",
+            feel = Property.FeelMode.disabled,
+            tooltip =
+                "System instructions inform how the model should respond."
+                    + " Use them to give the model context to understand the task, "
```
Comment (DenovVasil) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:75]: j

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/caller/GeminiCaller.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+public class GeminiCaller {
+
+  public static final String MIME_KEY = "mime";
+  public static final String URI_KEY = "uri";
+  public static final String TEXT_KEY = "text";
+
+  public Content generateContent(GeminiRequest geminiRequest) throws Exception {
+
+    GeminiRequestData requestData = geminiRequest.getInput();
+    List<Tool> tools = new ArrayList<>();
```
Comment (DenovVasil) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/caller/GeminiCaller.java:39]: d

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/mapper/PromptsMapper.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.List;
+import java.util.Map;
+
+public class PromptsMapper {
+
+  public static final String MIME_KEY = "mime";
+  public static final String URI_KEY = "uri";
+  public static final String TEXT_KEY = "text";
+
+  public Object[] map(List<Object> prompts) {
```
[the remark this answers is not in the document]
Comment (DenovVasil) [reply connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/mapper/PromptsMapper.java:30]: done

**Comment by johnBgood** [comment]
# QA information
## Test Environment
- Environment:
  - [ ] SaaS only
  - [ ] SM only
  - [ ] Both
- Connectors version: <TO_BE_REPLACED>

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*

File: connectors/google/google-gemini/src/main/resources/icon.svg
Comment (sbuettner) [inline connectors/google/google-gemini/src/main/resources/icon.svg:1]: @YanaSegal Is this the official gemini icon we want to use? Just curious.

Comment (YanaSegal) [reply connectors/google/google-gemini/src/main/resources/icon.svg:1]: i provided the official icon here https://github.com/camunda/product-hub/issues/2532#issuecomment-2459182338
if it uses this icon then yes :)

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+                    + "A lower temperature is good when you expect a true or correct response. \n"
+                    + "A temperature of 0 means the highest probability token is usually selected.\n"
+                    + "A higher temperature can lead to diverse or unexpected results. Some models have a higher temperature max to encourage more random responses.",
+            constraints =
+                @TemplateProperty.PropertyConstraints(
+                    pattern =
+                        @TemplateProperty.Pattern(
+                            value = "(^(([0-1]\\.[0-9])|([0-2]))$)|(^$)",
+                            message =
+                                "value must be in the range from 0 to 2 in increments of 0.1")))
```
Comment (johnBgood) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:270]: value => Value

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        @NotNull
+        ModelVersion model,
+    @FEEL
+        @TemplateProperty(
+            label = "Prompt",
+            group = "input",
+            description = "Insert prompt.",
+            feel = Property.FeelMode.required)
+        @NotNull
+        List<Object> prompts,
```
Comment (johnBgood) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:67]: I'm never comfortable being able to log the prompts. Do we have a decision on that? It affects all the AI connectors

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+                    equalsBoolean = TemplateProperty.EqualsBoolean.TRUE),
+            optional = true,
+            constraints =
+                @TemplateProperty.PropertyConstraints(
+                    pattern =
+                        @TemplateProperty.Pattern(
+                            value =
+                                "(^projects\\/.*\\/locations\\/.*\\/collections\\/.*\\/dataStores\\/.*$)",
+                            message =
+                                "value must match this template: projects/{}/locations/{}/collections/{}/dataStores/{}")))
```
Comment (johnBgood) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:121]: value => Value (check for other occurrences I might have missed some ^^)

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+                "Output token limit determines the maximum amount of text output from one prompt. "
+                    + "A token is approximately four characters.",
+            constraints =
+                @TemplateProperty.PropertyConstraints(
+                    pattern =
+                        @TemplateProperty.Pattern(
+                            value =
+                                "(^([1-9]|[1-9]\\d{1,2}|[1-7]\\d{3}|8(0[0-9]{2}|1[0-8][0-9]|19[0-2]))$)|(^$)",
+                            message =
+                                "value must be in the range from 1 to 8192 in increments of 1")))
```
Comment (johnBgood) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:286]: Value

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+                    + " Tokens are selected from most probable to least until the sum of their probabilities equals the top-p value."
+                    + " For example, if tokens A, B, and C have a probability of .3, .2, and .1 and the top-p value is .5, then the model will select either A or B as the next token (using temperature)."
+                    + " For the least variable results, set top-P to 0.",
+            constraints =
+                @TemplateProperty.PropertyConstraints(
+                    pattern =
+                        @TemplateProperty.Pattern(
+                            value = "(^((0\\.[0-9])|1|0)$)|(^$)",
+                            message =
+                                "value must be in the range from 0 to 1 in increments of 0.1")))
```
Comment (johnBgood) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:341]: Value

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+            condition =
+                @TemplateProperty.PropertyCondition(
+                    property = "input.safetySettings",
+                    equalsBoolean = TemplateProperty.EqualsBoolean.TRUE),
+            tooltip =
+                "You can adjust the likelihood of receiving a model response that could contain harmful content. "
+                    + "Content is blocked based on the probability that it's harmful."
+                    + "<a href=\"https://cloud.google.com/vertex-ai/docs/generative-ai/learn/responsible-ai?hl=en#safety_filters_and_attributes\" Learn more </a>",
+            optional = true)
+        BlockingDegree hateSpeach,
```
Comment (johnBgood) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:161]: => hateSpeech

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+ */
+package io.camunda.connector.gemini.model;
+
+import io.camunda.connector.feel.annotation.FEEL;
+import io.camunda.connector.generator.dsl.Property;
+import io.camunda.connector.generator.java.annotation.TemplateProperty;
+import jakarta.validation.constraints.NotNull;
+import java.util.List;
+
+public record GeminiRequestData(
```
Comment (johnBgood) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:15]: has the element template been reviewed by Yana? Just want to be sure as there's a lot of properties there :D

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/mapper/PromptsMapper.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    switch (map.size()) {
+      case 1 -> {
+        validateEntry(map, TEXT_KEY);
+        return map.get(TEXT_KEY);
+      }
+      case 2 -> {
+        validateEntry(map, MIME_KEY, URI_KEY);
+        return PartMaker.fromMimeTypeAndData(map.get(MIME_KEY), map.get(URI_KEY));
+      }
+      default -> throw createInvalidPromptException(String.format(INVALID_PROMPT_MSG_FORMAT, map));
```
Comment (johnBgood) [inline connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/mapper/PromptsMapper.java:65]: maybe can we add more context to this error message, if we know what's wrong. Here apparently we need a size of 1 or 2. Same for other occurrences

**Review by johnBgood — COMMENTED** [review]
Some minor questions, great work :)

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/mapper/PromptsMapper.java (hunk 12 above)
Comment (DenovVasil) [reply connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/mapper/PromptsMapper.java:65]: I don't think it's necessary, the documentation fully describes what the format of the `feel` expression for prompts should be

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java (hunk 11 above)
Comment (DenovVasil) [reply connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:15]: This template fully displays the Gemini interface from Google Cloud, Christian looked at it earlier and he had no comments

File: connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+            tooltip =
+                "Top-K specifies the number of candidate tokens when the model is selecting an output token. "
+                    + "Use a lower value for less random responses and a higher value for more random responses.",
+            constraints =
+                @TemplateProperty.PropertyConstraints(
+                    pattern =
+                        @TemplateProperty.Pattern(
+                            value = "(^([1-9]|[1-3][0-9]|40)$)(^$)",
+                            message = "value must be an integer between 1 and 40")))
+        float topK,
```
[the remark this answers is not in the document]
Comment (johnBgood) [reply connectors/google/google-gemini/src/main/java/io/camunda/connector/gemini/model/GeminiRequestData.java:324]: I don't see any `.` in the regex, below you have a different regex that includes dots

**Review by johnBgood — APPROVED** [review]
